File: src/main/java/gargoyle/rpycg/service/Validator.java
package gargoyle.rpycg.service;

import javafx.beans.property.Property;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.util.Callback;

import java.util.*;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public final class Validator {
    private final Map<Property<?>, Consumer<Set<String>>> listeners = new LinkedHashMap<>(2);
    private final SimpleBooleanProperty valid = new SimpleBooleanProperty(true);
    private final Map<Property<?>, Set<Callback<Object, String>>> validators = new LinkedHashMap<>(2);

    @SuppressWarnings("unchecked")
    public <T> void addValidator(final Property<T> property,
                                 final Callback<T, String> validator, final Consumer<Set<String>> listener) {
        if (!validators.containsKey(property)) {
            listeners.put(property, listener);
            validators.put(property, new HashSet<>(2));
            property.addListener((observable, oldValue, newValue) -> {
                final Set<String> errors = validators.get(property).stream()
                        .map(v -> v.call(property.getValue()))
                        .filter(Objects::nonNull)
                        .collect(Collectors.toSet());
                errors.removeIf(Objects::isNull);
                validate();
                listener.accept(errors);
            });
        }
        validators.get(property).add((Callback<Object, String>) validator);
    }

    public boolean isValid() {
        return valid.getValue();
    }

    public SimpleBooleanProperty validProperty() {
        return valid;
    }

    public Map<Property<?>, Set<String>> validate() {
        final Map<Property<?>, Set<String>> errors = validators.entrySet().stream()
                .collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().stream()
                        .map(predicate -> predicate.call(entry.getKey().getValue()))
                        .filter(Objects::nonNull)
                        .collect(Collectors.toSet())));
        listeners.forEach((property, listConsumer) -> listConsumer.accept(errors.get(property)));
        errors.entrySet().removeIf(entries -> entries.getValue().isEmpty());
        valid.setValue(errors.isEmpty());
        return errors;
    }
}

File: app/src/main/kotlin/gargoyle/rpycg/service/Validator.kt
package gargoyle.rpycg.service

import javafx.beans.property.Property
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.value.ObservableValue

@Suppress("unused")
class Validator {
    private val listeners: MutableMap<Property<*>, (Set<String>) -> Unit> = LinkedHashMap(2)
    private val valid = SimpleBooleanProperty(true)
    private val validators: MutableMap<Property<*>, MutableSet<(Any) -> String>> = LinkedHashMap(2)

    @Suppress("UNCHECKED_CAST")
    fun <T : Any> addValidator(property: Property<T>, validator: (T) -> String?, listener: (Set<String>) -> Unit) {
        if (!validators.containsKey(property)) {
            listeners[property] = listener
            validators[property] = mutableSetOf()
            property.addListener { _: ObservableValue<out T>, _: T, _: T ->
                val errors = validators[property]!!
                    .map { v: (Any) -> String -> v(property.value!!) }
                    .filter { obj: String? -> null != (obj) }
                validate()
                listener(errors.toSet())
            }
        }
        validators[property]!!.add(validator as (Any) -> String)
    }

    fun isValid(): Boolean = valid.value

    fun validProperty(): SimpleBooleanProperty = valid

    fun validate(): Map<Property<*>, Set<String>> {
        val errors: MutableMap<Property<*>, MutableSet<String>> =
            validators.mapValues { entry: Map.Entry<Property<*>, MutableSet<(Any) -> String>> ->
                entry.value.map { function -> function(entry.key.value) }.toMutableSet()
            }.toMutableMap()
        listeners.forEach { (property: Property<*>, listConsumer: (Set<String>) -> Unit) ->
            listConsumer(errors[property]!!)
        }
        errors.entries.removeIf { (_, value): Map.Entry<Property<*>, Set<String>> -> value.isEmpty() }
        valid.value = errors.isEmpty()
        return errors
    }
}
